// @axols/webai-js/dist/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  ErrorCategory: () => ErrorCategory,
  WebAI: () => WebAI,
  checkIsModelDownloaded: () => checkIsModelDownloaded,
  checkIsWebGPUAvailable: () => checkIsWebGPUAvailable,
  checkStorageQuota: () => checkStorageQuota,
  clearModelCache: () => clearModelCache,
  createAuthError: () => createAuthError,
  createError: () => createError,
  createNetworkError: () => createNetworkError,
  createWorkerError: () => createWorkerError,
  identifyError: () => identifyError,
  isAuthError: () => isAuthError,
  isErrorOf: () => isErrorOf,
  isLocalPath: () => isLocalPath,
  isNetworkError: () => isNetworkError,
  isWorkerError: () => isWorkerError
});
module.exports = __toCommonJS(index_exports);

// src/utils/errors.ts
var ErrorCategory = /* @__PURE__ */ ((ErrorCategory2) => {
  ErrorCategory2["AUTHENTICATION"] = "AuthError";
  ErrorCategory2["VALIDATION"] = "ValidationError";
  ErrorCategory2["NETWORK"] = "NetworkError";
  ErrorCategory2["SYSTEM"] = "SystemError";
  ErrorCategory2["WORKER"] = "WorkerError";
  ErrorCategory2["UNEXPECTED"] = "UnexpectedError";
  ErrorCategory2["RESOURCE"] = "ResourceError";
  return ErrorCategory2;
})(ErrorCategory || {});
function createError(message, category, type, details = {}) {
  const formattedMessage = `[${category}:${type}] ${message}`;
  const error = new Error(formattedMessage);
  Object.assign(error, { details });
  return error;
}
function identifyError(error) {
  if (!error || !error.message) return null;
  const match = error.message.match(/^\[(.*?):(.*?)\]/);
  if (match) {
    return {
      category: match[1],
      type: match[2]
    };
  }
  return null;
}
function createAuthError(message, type, details = {}) {
  return createError(message, "AuthError" /* AUTHENTICATION */, type, details);
}
function createWorkerError(message, type, details = {}) {
  return createError(message, "WorkerError" /* WORKER */, type, details);
}
function createNetworkError(message, type, details = {}) {
  return createError(message, "NetworkError" /* NETWORK */, type, details);
}
function isErrorOf(error, category, type) {
  const errorInfo = identifyError(error);
  if (!errorInfo) return false;
  return (!category || errorInfo.category === category) && (!type || errorInfo.type === type);
}
function isAuthError(error) {
  const errorInfo = identifyError(error);
  return errorInfo?.category === "AuthError" /* AUTHENTICATION */;
}
function isWorkerError(error) {
  const errorInfo = identifyError(error);
  return errorInfo?.category === "WorkerError" /* WORKER */;
}
function isNetworkError(error) {
  const errorInfo = identifyError(error);
  return errorInfo?.category === "NetworkError" /* NETWORK */;
}

// src/utils/utils.ts
async function checkIsWebGPUAvailable() {
  try {
    if (!navigator || !navigator.gpu) {
      return false;
    }
    const adapter = await navigator.gpu.requestAdapter();
    return adapter !== null;
  } catch (error) {
    console.log("Error checking WebGPU availability:", error);
    return false;
  }
}
async function checkStorageQuota() {
  try {
    if (navigator.storage && navigator.storage.estimate) {
      const estimate = await navigator.storage.estimate();
      return estimate;
    } else {
      console.log("Storage API not supported in this browser");
      return null;
    }
  } catch (error) {
    console.error("Error checking storage quota:", error);
    return null;
  }
}
async function checkIsModelDownloaded(precision, modelKeys, modelId) {
  const cache = await caches.open("transformers-cache");
  const keys = (await cache.keys()).map((key) => key.url);
  if (keys.length === 0) {
    return false;
  }
  const isDownloaded = modelKeys.every(
    (modelKey) => keys.some((key) => key.includes(modelKey) && key.includes("/" + modelId + "/"))
  );
  if (!isDownloaded) {
    return false;
  }
  return isDownloaded;
}
async function clearModelCache(modelId, modelKeys) {
  if (!modelId) {
    throw createError(
      "Model ID is required to clear cache",
      "ResourceError" /* RESOURCE */,
      "missing_parameter"
    );
  }
  try {
    const cache = await caches.open("transformers-cache");
    const cacheKeys = await cache.keys();
    const keys = cacheKeys.map((key) => key.url);
    const allModelKeys = keys.filter((key) => key.includes("/" + modelId + "/"));
    await Promise.allSettled(allModelKeys.map((key) => cache.delete(new Request(key))));
  } catch (error) {
    const errorMessage = error instanceof Error ? error.message : "Unknown error";
    throw createError(
      `Failed to clear model cache: ${errorMessage}`,
      "ResourceError" /* RESOURCE */,
      "cache_operation_failed"
    );
  }
}
function isLocalPath(path) {
  if (path.includes("://")) {
    return false;
  }
  if (path.startsWith("./") || path.startsWith("../")) {
    return true;
  }
  if (path.startsWith("/")) {
    return true;
  }
  if (/^[a-zA-Z]:[/\\]/.test(path)) {
    return true;
  }
  if (path.startsWith("\\\\")) {
    return true;
  }
  return true;
}

// src/utils/constants.ts
var VERSION = "0.0.7";
var REMOTE_WORKER_ENDPOINT = "https://assets.axolsai.com";
var API_ENDPOINT = "https://api.webai-js.com";

// src/utils/download-logs.ts
async function logStartDownload(modelId, precision, price = 0, isDev = false, clientIdentifier, authToken) {
  try {
    const headers = {
      "Content-Type": "application/json"
    };
    if (authToken) {
      headers["Authorization"] = `Bearer ${authToken}`;
    }
    const response = await fetch(`${API_ENDPOINT}/download/start`, {
      method: "POST",
      headers,
      body: JSON.stringify({
        modelId,
        precision,
        price,
        type: isDev ? "dev" : "prod",
        clientIdentifier
      })
    });
    if (!response.ok) {
      const errorData = await response.json().catch(() => ({}));
      throw createError(
        `Failed to log download start: ${errorData.error || response.statusText}`,
        "NetworkError" /* NETWORK */,
        "server_error"
      );
    }
    const { downloadId } = await response.json();
    if (!downloadId) {
      throw createError(
        "Server did not return a downloadId",
        "NetworkError" /* NETWORK */,
        "server_error"
      );
    }
    return downloadId;
  } catch (error) {
    const errorMessage = error instanceof Error ? error.message : "Unknown error";
    throw createError(
      `Failed to log start download: ${errorMessage}`,
      "NetworkError" /* NETWORK */,
      "request_failed"
    );
  }
}
async function logEndDownload(downloadId) {
  try {
    const response = await fetch(`${API_ENDPOINT}/download/end`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json"
      },
      body: JSON.stringify({
        downloadId
      })
    });
    if (!response.ok) {
      const errorData = await response.json().catch(() => ({}));
      throw createError(
        `Failed to log download end: ${errorData.error || response.statusText}`,
        "NetworkError" /* NETWORK */,
        "server_error"
      );
    }
  } catch (error) {
    console.error("Failed to log download completion:", error);
  }
}
async function logFailedDownload(downloadId) {
  try {
    const response = await fetch(`${API_ENDPOINT}/download/failed`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json"
      },
      body: JSON.stringify({
        downloadId
      })
    });
    if (!response.ok) {
      const errorData = await response.json().catch(() => ({}));
      throw createError(
        `Failed to log download failure: ${errorData.error || response.statusText}`,
        "NetworkError" /* NETWORK */,
        "server_error"
      );
    }
  } catch (error) {
    console.error("Failed to log download failure:", error);
  }
}

// src/managers/webai-model-manager.ts
var ModelManager = class {
  _modelId;
  _isDev = false;
  _workerManager = null;
  _onAuth;
  _authRetryOptions;
  _mode = null;
  _precision = null;
  _device = null;
  _modelSupportedModes = null;
  _modelSupportedPrecisions = null;
  _supportedPrecisionsDevicesMap = null;
  _doesSupportStreamGeneration = null;
  _externalInterrupt = null;
  constructor(modelId, workerManager = null, isDev = false, onAuth, authRetryOptions) {
    this._modelId = modelId;
    this._workerManager = workerManager;
    this._isDev = isDev;
    this._onAuth = onAuth;
    this._authRetryOptions = authRetryOptions;
  }
  /**
   * Attempts to get an auth token with retry logic
   */
  async getAuthToken() {
    if (!this._onAuth) {
      return void 0;
    }
    const maxRetries = this._authRetryOptions?.maxRetries ?? 3;
    const baseRetryInterval = this._authRetryOptions?.retryInterval ?? 2e3;
    const exponentialBackoff = this._authRetryOptions?.exponentialBackoff ?? true;
    for (let attempt = 0; attempt <= maxRetries; attempt++) {
      try {
        const token = await this._onAuth();
        if (token && typeof token === "string") {
          return token;
        }
        throw new Error("onAuth callback did not return a valid token string");
      } catch (error) {
        console.error(`Auth attempt ${attempt + 1} failed:`, error);
        if (attempt < maxRetries) {
          const waitTime = exponentialBackoff ? baseRetryInterval * Math.pow(2, attempt) : baseRetryInterval;
          console.log(`Retrying in ${waitTime}ms...`);
          await new Promise((resolve) => setTimeout(resolve, waitTime));
        } else {
          console.error("All auth retry attempts failed");
          throw createWorkerError(
            `Failed to get auth token after ${maxRetries + 1} attempts`,
            "initialization_failed"
          );
        }
      }
    }
    return void 0;
  }
  async checkModelSupports(worker) {
    if (!worker) {
      throw createWorkerError(
        "Worker is not initialized",
        "initialization_failed"
      );
    }
    return new Promise((resolve, reject) => {
      const cleanup = () => {
        worker.removeEventListener("message", handleMessage);
        worker.removeEventListener("error", handleError);
      };
      const handleMessage = (event) => {
        if (event.data.type === "error") {
          cleanup();
          reject(createWorkerError(
            event.data.data.message || "Error checking model support",
            "runtime_error"
          ));
          return;
        }
        if (event.data.type === "checkModelSupports") {
          cleanup();
          const {
            supportedModes,
            supportedPrecisions,
            supportedPrecisionsDevicesMap,
            doesSupportStreamGeneration,
            externalInterrupt
          } = event.data.data;
          this._modelSupportedModes = supportedModes;
          this._modelSupportedPrecisions = supportedPrecisions;
          this._supportedPrecisionsDevicesMap = supportedPrecisionsDevicesMap;
          this._doesSupportStreamGeneration = doesSupportStreamGeneration;
          this._externalInterrupt = externalInterrupt;
          resolve();
        }
      };
      const handleError = (error) => {
        cleanup();
        reject(createWorkerError(
          `Worker error during model support check: ${error.message}`,
          "runtime_error"
        ));
      };
      worker.addEventListener("message", handleMessage);
      worker.addEventListener("error", handleError);
      try {
        worker.postMessage({ type: "checkModelSupports" });
      } catch (error) {
        cleanup();
        reject(createWorkerError(
          error instanceof Error ? `Failed to check model supports: ${error.message}` : "Failed to check model supports: Unknown error",
          "runtime_error"
        ));
      }
    });
  }
  async sendInitMessage(worker, precision, device) {
    if (!worker) {
      throw createWorkerError(
        "Worker is not initialized",
        "initialization_failed"
      );
    }
    return new Promise((resolve, reject) => {
      const cleanup = () => {
        worker.removeEventListener("message", handleMessage);
        worker.removeEventListener("error", handleError);
      };
      const handleMessage = (event) => {
        if (event.data.type === "error") {
          cleanup();
          reject(createWorkerError(
            event.data.data.message || "Unknown error",
            "runtime_error"
          ));
          return;
        }
        if (event.data.type === "init") {
          cleanup();
          if (event.data.data.status === "success") {
            this._device = device;
            this._precision = precision;
            resolve(true);
          } else {
            reject(createWorkerError(
              `Failed to initialize with device ${device} and precision ${precision}`,
              "runtime_error"
            ));
          }
        }
      };
      const handleError = (error) => {
        cleanup();
        reject(createWorkerError(
          `Worker error during initialization: ${error.message}`,
          "runtime_error"
        ));
      };
      worker.addEventListener("message", handleMessage);
      worker.addEventListener("error", handleError);
      try {
        worker.postMessage({
          type: "init",
          data: { precision, device }
        });
      } catch (error) {
        cleanup();
        reject(createWorkerError(
          error instanceof Error ? `Failed to send init request: ${error.message}` : "Failed to send init request: Unknown error",
          "runtime_error"
        ));
      }
    });
  }
  async downloadModel(worker, precision, downloadProgressCallback, callbackThrottle = 500) {
    if (!worker) {
      throw createWorkerError(
        "Worker is not initialized",
        "initialization_failed"
      );
    }
    const modelKeys = this._supportedPrecisionsDevicesMap?.[precision]?.modelKeys || [];
    const isModelDownloaded = await checkIsModelDownloaded(
      precision,
      modelKeys,
      this._modelId
    );
    if (isModelDownloaded) {
      if (downloadProgressCallback) {
        const actualSize = this._supportedPrecisionsDevicesMap?.[precision]?.size || 0;
        downloadProgressCallback({
          file: `${precision} models`,
          name: this._modelId,
          loaded: actualSize,
          total: actualSize,
          status: "downloaded",
          progress: 100
        });
      }
      return true;
    }
    const price = this._supportedPrecisionsDevicesMap?.[precision]?.price || 0;
    let authToken;
    try {
      authToken = await this.getAuthToken();
    } catch (error) {
      throw createWorkerError(
        `Failed to get auth token: ${error instanceof Error ? error.message : "Unknown error"}`,
        "initialization_failed"
      );
    }
    let downloadId = null;
    try {
      downloadId = await logStartDownload(
        this._modelId,
        precision,
        price,
        this._isDev,
        void 0,
        // clientIdentifier will be extracted from token on server
        authToken
      );
      console.log(`Download started with ID: ${downloadId}`);
    } catch (error) {
      throw createWorkerError(
        `Failed to log download start: ${error instanceof Error ? error.message : "Unknown error"}. Download aborted.`,
        "initialization_failed"
      );
    }
    return new Promise((resolve, reject) => {
      let lastProgressUpdate = 0;
      const fileProgress = {};
      const cleanup = () => {
        worker.removeEventListener("message", handleMessage);
        worker.removeEventListener("error", handleError);
      };
      const handleMessage = async (event) => {
        if (event.data.type === "error") {
          cleanup();
          if (downloadId) {
            await logFailedDownload(downloadId);
          }
          reject(createWorkerError(
            event.data.data.message || "Download error",
            "runtime_error"
          ));
          return;
        }
        if (event.data.type === "downloadProgress") {
          const progressData = event.data.data;
          const filename = progressData.file;
          if (!filename) return;
          const matchedKey = modelKeys.find((key) => filename.includes(key));
          const now = Date.now();
          if (matchedKey && now - lastProgressUpdate > callbackThrottle) {
            fileProgress[matchedKey] = progressData;
            const allKeysPresent = modelKeys.every((key) => fileProgress[key] !== void 0);
            if (allKeysPresent && downloadProgressCallback) {
              let totalLoaded = 0;
              let totalSize = 0;
              modelKeys.forEach((key) => {
                totalLoaded += fileProgress[key].loaded;
                totalSize += fileProgress[key].total;
              });
              const progress = totalLoaded / totalSize * 100;
              if (progress >= 0 && progress <= 100) {
                downloadProgressCallback({
                  file: `${precision} models`,
                  name: this._modelId,
                  loaded: totalLoaded,
                  total: totalSize,
                  status: "downloading",
                  progress
                });
              }
            }
            lastProgressUpdate = now;
          }
          return;
        }
        if (event.data.type === "download" && event.data.data.status === "success") {
          if (downloadProgressCallback) {
            const actualSize = this._supportedPrecisionsDevicesMap?.[precision]?.size || 0;
            downloadProgressCallback({
              file: `${precision} models`,
              name: this._modelId,
              loaded: actualSize,
              total: actualSize,
              status: "downloaded",
              progress: 100
            });
          }
          cleanup();
          if (downloadId) {
            await logEndDownload(downloadId);
          }
          if (this._workerManager) {
            try {
              console.log("Reloading worker after successful download...");
              await this._workerManager.reloadWorker();
              if (!this._workerManager.worker) {
                throw createWorkerError(
                  "Worker is null after reload",
                  "initialization_failed"
                );
              }
              await this.checkModelSupports(this._workerManager.worker);
            } catch (error) {
              console.error("Failed to reload worker after download:", error);
              if (downloadId) {
                await logFailedDownload(downloadId);
              }
              reject(createWorkerError(
                "Failed to reload worker after download",
                "initialization_failed"
              ));
              return;
            }
          }
          resolve(true);
        }
      };
      const handleError = async (error) => {
        cleanup();
        if (downloadId) {
          await logFailedDownload(downloadId);
        }
        reject(createWorkerError(
          `Worker error during download: ${error.message}`,
          "runtime_error"
        ));
      };
      worker.addEventListener("message", handleMessage);
      worker.addEventListener("error", handleError);
      try {
        worker.postMessage({ type: "download", data: { precision } });
      } catch (error) {
        cleanup();
        if (downloadId) {
          logFailedDownload(downloadId).catch(console.error);
        }
        reject(createWorkerError(
          error instanceof Error ? `Failed to send download request: ${error.message}` : "Failed to send download request: Unknown error",
          "runtime_error"
        ));
      }
    });
  }
  validateInitConfig(mode, precision, device) {
    if (!this._modelSupportedModes?.includes(mode)) {
      throw createWorkerError(
        `Mode '${mode}' is not supported for this model.`,
        "initialization_failed"
      );
    }
    if (mode === "webai") {
      if (!device || !precision) {
        throw createWorkerError(
          "For 'webai' mode, both device and precision must be specified.",
          "initialization_failed"
        );
      }
      if (!this._modelSupportedPrecisions?.includes(precision)) {
        throw createWorkerError(
          `Precision '${precision}' is not supported for this model. Supported precisions are: ${this._modelSupportedPrecisions?.join(", ")}`,
          "initialization_failed"
        );
      }
      const supportedDevicesForPrecision = this._supportedPrecisionsDevicesMap?.[precision]?.supportedDevices || [];
      if (!supportedDevicesForPrecision.includes(device)) {
        throw createWorkerError(
          `Device '${device}' is not supported for precision '${precision}'. Supported devices for this precision are: ${supportedDevicesForPrecision.join(", ")}`,
          "initialization_failed"
        );
      }
    }
  }
  validateStorageAndWebGPU(device, precision, isWebGPUAvailable, storageQuota) {
    if (device === "webgpu" && !isWebGPUAvailable) {
      throw createWorkerError(
        "WebGPU is not available in your current browser. We recommend using the latest version of Chrome. You can find a list of supported browsers at https://caniuse.com/webgpu",
        "initialization_failed"
      );
    }
    const modelSize = this._supportedPrecisionsDevicesMap?.[precision]?.size;
    if (!modelSize) {
      throw createWorkerError(
        `Model size information not available for precision '${precision}'.`,
        "initialization_failed"
      );
    }
    const leftStorage = (storageQuota?.quota ?? 0) - (storageQuota?.usage ?? 0);
    if (modelSize > leftStorage) {
      throw createWorkerError(
        `Not enough storage to download the model. Required: ${modelSize} bytes, Available: ${leftStorage} bytes.`,
        "initialization_failed"
      );
    }
  }
  async initModel({
    mode = "auto",
    precision,
    device,
    priorities,
    onDownloadProgress,
    callbackThrottle = 1e3,
    worker,
    isWebGPUAvailable,
    storageQuota
  }) {
    if (!this._modelSupportedModes || !this._supportedPrecisionsDevicesMap || !this._modelSupportedPrecisions) {
      throw createWorkerError(
        "Model support information is not available. Make sure to create the instance properly.",
        "initialization_failed"
      );
    }
    if (mode !== "auto") {
      this.validateInitConfig(mode, precision, device);
      if (mode === "cloud") {
        this._mode = "cloud";
        return true;
      }
      if (mode === "webai" && device && precision) {
        this.validateStorageAndWebGPU(device, precision, isWebGPUAvailable, storageQuota);
        this._mode = "webai";
        const result = await this.#initWebAI(
          worker,
          precision,
          device,
          onDownloadProgress,
          callbackThrottle
        );
        if (!result) {
          throw createWorkerError("Failed to initialize model", "initialization_failed");
        }
        return true;
      }
      throw createWorkerError(`Unsupported mode: ${mode}`, "initialization_failed");
    }
    const defaultPriorities = this.getDefaultPriorities();
    const priorityList = priorities?.length ? priorities : defaultPriorities;
    for (const config of priorityList) {
      if (await this.tryConfiguration(
        config,
        worker,
        isWebGPUAvailable,
        storageQuota,
        onDownloadProgress,
        callbackThrottle
      )) {
        return true;
      }
    }
    throw createWorkerError(
      "Could not run the model, none of the configurations (mode, precision, device) specified in the priorities list works. Please verify the model requirements and the device you are running on.",
      "initialization_failed"
    );
  }
  getDefaultPriorities() {
    return [
      { mode: "webai", precision: "q4", device: "webgpu" },
      { mode: "webai", precision: "q4f16", device: "webgpu" },
      { mode: "webai", precision: "bnb4", device: "webgpu" },
      { mode: "webai", precision: "q8", device: "webgpu" },
      { mode: "webai", precision: "int8", device: "webgpu" },
      { mode: "webai", precision: "uint8", device: "webgpu" },
      { mode: "webai", precision: "fp16", device: "webgpu" },
      { mode: "webai", precision: "fp32", device: "webgpu" },
      { mode: "webai", precision: "q4", device: "wasm" },
      { mode: "webai", precision: "q4f16", device: "wasm" },
      { mode: "webai", precision: "bnb4", device: "wasm" },
      { mode: "webai", precision: "q8", device: "wasm" },
      { mode: "webai", precision: "int8", device: "wasm" },
      { mode: "webai", precision: "uint8", device: "wasm" },
      { mode: "webai", precision: "fp16", device: "wasm" },
      { mode: "webai", precision: "fp32", device: "wasm" },
      { mode: "cloud", precision: "", device: "" }
    ];
  }
  async tryConfiguration(config, worker, isWebGPUAvailable, storageQuota, onDownloadProgress, callbackThrottle) {
    const currentMode = config.mode;
    if (!this._modelSupportedModes?.includes(currentMode)) {
      return false;
    }
    if (currentMode === "cloud") {
      this._mode = "cloud";
      return true;
    }
    if (currentMode === "webai") {
      const precision = config.precision;
      const device = config.device;
      if (!this._modelSupportedPrecisions?.includes(precision)) {
        return false;
      }
      const supportedDevices = this._supportedPrecisionsDevicesMap?.[precision]?.supportedDevices || [];
      if (!supportedDevices.includes(device)) {
        return false;
      }
      if (device === "webgpu" && !isWebGPUAvailable) {
        return false;
      }
      const modelSize = this._supportedPrecisionsDevicesMap?.[precision]?.size;
      if (!modelSize) {
        return false;
      }
      const leftStorage = (storageQuota?.quota ?? 0) - (storageQuota?.usage ?? 0);
      if (modelSize > leftStorage) {
        return false;
      }
      try {
        console.log("Trying configuration:", precision, device);
        const result = await this.#initWebAI(
          worker,
          precision,
          device,
          onDownloadProgress,
          callbackThrottle
        );
        if (result) {
          this._mode = "webai";
          return true;
        }
      } catch (error) {
        console.warn(`Failed to initialize with ${precision}/${device}:`, error);
      }
    }
    return false;
  }
  async #initWebAI(worker, precision, device, onDownloadProgress, callbackThrottle = 1e3) {
    const modelKeys = this._supportedPrecisionsDevicesMap?.[precision]?.modelKeys || [];
    const isModelDownloaded = await checkIsModelDownloaded(precision, modelKeys, this._modelId);
    if (isModelDownloaded) {
      await this.sendInitMessage(worker, precision, device);
      return true;
    }
    const downloadSuccess = await this.downloadModel(
      worker,
      precision,
      onDownloadProgress,
      callbackThrottle
    );
    if (!downloadSuccess) {
      throw createWorkerError("Failed to download model", "initialization_failed");
    }
    const freshWorker = this._workerManager?.worker;
    if (!freshWorker) {
      throw createWorkerError(
        "Worker is null after download and reload",
        "initialization_failed"
      );
    }
    await this.sendInitMessage(freshWorker, precision, device);
    return true;
  }
  async reinitializeAfterInterrupt(worker) {
    if (!worker || !this._mode || !this._precision || !this._device) {
      throw createWorkerError(
        "Cannot reinitialize: missing required state (mode, precision, or device)",
        "initialization_failed"
      );
    }
    if (this._mode !== "webai") {
      return;
    }
    try {
      const modelKeys = this._supportedPrecisionsDevicesMap?.[this._precision]?.modelKeys;
      if (!modelKeys) {
        throw createWorkerError(
          `Model does not support precision ${this._precision}`,
          "initialization_failed"
        );
      }
      const isDownloaded = await checkIsModelDownloaded(
        this._precision,
        modelKeys,
        this._modelId
      );
      if (!isDownloaded) {
        throw createWorkerError(
          "Model is no longer downloaded - full reinitialization required",
          "initialization_failed"
        );
      }
      await this.checkModelSupports(worker);
      await this.sendInitMessage(worker, this._precision, this._device);
      console.log(`Model reinitialized after interrupt: ${this._precision}/${this._device}`);
    } catch (error) {
      console.error("Failed to reinitialize model after interrupt:", error);
      this._mode = null;
      this._precision = null;
      this._device = null;
      throw error;
    }
  }
  async clearCache() {
    return clearModelCache(this._modelId);
  }
  async deletePrecisionSpecificModel(precision) {
    if (!this._supportedPrecisionsDevicesMap?.[precision]) {
      console.warn(`Precision ${precision} is not supported for model ${this._modelId}`);
      return;
    }
    const modelKeys = this._supportedPrecisionsDevicesMap[precision].modelKeys;
    const downloadedPrecisions = [];
    if (this._modelSupportedPrecisions) {
      for (const supportedPrecision of this._modelSupportedPrecisions) {
        if (supportedPrecision === precision) continue;
        const otherModelKeys = this._supportedPrecisionsDevicesMap?.[supportedPrecision]?.modelKeys;
        if (otherModelKeys) {
          const isDownloaded = await checkIsModelDownloaded(
            supportedPrecision,
            otherModelKeys,
            this._modelId
          );
          if (isDownloaded) {
            downloadedPrecisions.push(supportedPrecision);
          }
        }
      }
    }
    if (downloadedPrecisions.length === 0) {
      console.log(`Deleting all cached files for model ${this._modelId} (only precision: ${precision})`);
      await this.clearCache();
      return;
    }
    console.log(`Deleting precision-specific files for ${this._modelId}, precision: ${precision}`);
    console.log(`Keeping other precisions: ${downloadedPrecisions.join(", ")}`);
    const cache = await caches.open("transformers-cache");
    for (const modelKey of modelKeys) {
      try {
        const cacheKeys = await cache.keys();
        for (const request of cacheKeys) {
          const url = request.url;
          if (url.includes(this._modelId) && url.includes(modelKey)) {
            await cache.delete(request);
          }
        }
      } catch (error) {
        console.warn(`Failed to delete cached files for model key ${modelKey}:`, error);
      }
    }
    console.log(`Successfully deleted model files for precision: ${precision}`);
  }
  // Getters
  get mode() {
    return this._mode;
  }
  get precision() {
    return this._precision;
  }
  get device() {
    return this._device;
  }
  get modelSupportedModes() {
    return this._modelSupportedModes;
  }
  get modelSupportedPrecisions() {
    return this._modelSupportedPrecisions;
  }
  get modelSupportedPrecisionsDevicesMap() {
    return this._supportedPrecisionsDevicesMap;
  }
  get doesSupportStreamGeneration() {
    return this._doesSupportStreamGeneration;
  }
  get externalInterrupt() {
    return this._externalInterrupt;
  }
};

// src/managers/webai-queue-manager.ts
var QueueManager = class {
  _isGenerating = false;
  _queue = [];
  _currentExecution = null;
  // Track current execution
  constructor() {
  }
  async enqueueGenerate(data, worker, mode) {
    return new Promise((resolve, reject) => {
      this._queue.push({
        type: "generate",
        data,
        resolve,
        reject
      });
      if (!this._isGenerating) {
        this.processNextInQueue(worker, mode);
      }
    });
  }
  async enqueueGenerateStream(data, worker, mode) {
    return new Promise((resolve, reject) => {
      this._queue.push({
        type: "generateStream",
        data,
        resolve,
        reject
      });
      if (!this._isGenerating) {
        this.processNextInQueue(worker, mode);
      }
    });
  }
  /**
   * Handle forced interruption (for WASM worker reload)
   * This resolves the current generation with an interruption result
   */
  handleForcedInterruption() {
    if (this._currentExecution && this._isGenerating) {
      console.log("WebAI generation is manually interrupted (forced).");
      this._currentExecution.resolve(void 0);
      this._currentExecution = null;
    }
    this._isGenerating = false;
  }
  // Process the next item in the queue
  processNextInQueue(worker, mode) {
    if (this._queue.length === 0) {
      this._isGenerating = false;
      this._currentExecution = null;
      return;
    }
    this._isGenerating = true;
    const nextItem = this._queue.shift();
    const isStream = nextItem.type === "generateStream";
    const { onStream, ...dataWithoutOnStream } = nextItem.data;
    const context = {
      data: isStream ? dataWithoutOnStream : nextItem.data,
      resolve: nextItem.resolve,
      reject: nextItem.reject,
      worker,
      mode,
      isStream,
      onStream: isStream ? onStream : void 0
    };
    this._executeGeneration(context);
  }
  // Unified execution method for both generate and generateStream
  _executeGeneration(context) {
    const { data, resolve, reject, worker, mode, isStream, onStream } = context;
    this._currentExecution = context;
    if (!worker) {
      reject(createWorkerError(
        "Worker is not initialized",
        "initialization_failed"
      ));
      this._currentExecution = null;
      this.processNextInQueue(worker, mode);
      return;
    }
    if (!mode) {
      reject(createWorkerError(
        "WebAI is not initialized. Call init() before generate()",
        "initialization_failed"
      ));
      this._currentExecution = null;
      this.processNextInQueue(worker, mode);
      return;
    }
    const cleanup = () => {
      worker?.removeEventListener("message", handleMessage);
      worker?.removeEventListener("error", handleError);
      this._currentExecution = null;
    };
    const handleMessage = (event) => {
      if (event.data.type === "error") {
        const errorMessage = isStream ? "Stream generation error" : "Generation error";
        console.log(`WebAI ${isStream ? "stream " : ""}generation error:`, event.data.data.message);
        cleanup();
        reject(createWorkerError(
          event.data.data.message || errorMessage,
          "runtime_error"
        ));
        this.processNextInQueue(worker, mode);
        return;
      }
      if (event.data.type === "interrupted") {
        console.log(`WebAI generation is manually interrupted.`);
        cleanup();
        resolve(void 0);
        this.processNextInQueue(worker, mode);
        return;
      }
      if (event.data.type === "stream" && isStream && onStream) {
        onStream(event.data.data);
        return;
      }
      if (event.data.type === "generated") {
        if (isStream) {
          console.log("stream ****", event.data.data.result);
        }
        cleanup();
        resolve(event.data.data.result);
        this.processNextInQueue(worker, mode);
        return;
      }
      if (event.data.type === "generateError") {
        cleanup();
        reject(createWorkerError(
          `Generation error: ${event.data.error}`,
          "runtime_error"
        ));
        this.processNextInQueue(worker, mode);
        return;
      }
    };
    const handleError = (error) => {
      const errorMessage = isStream ? `Worker error during stream generation: ${error.message}` : `Worker error during generation: ${error.message}`;
      cleanup();
      reject(createWorkerError(errorMessage, "runtime_error"));
      this.processNextInQueue(worker, mode);
    };
    worker.addEventListener("message", handleMessage);
    worker.addEventListener("error", handleError);
    try {
      worker.postMessage({
        type: isStream ? "generateStream" : "generate",
        data
      });
    } catch (error) {
      const errorMessage = isStream ? "Failed to send stream generation request" : "Failed to send generation request";
      cleanup();
      reject(createWorkerError(
        error instanceof Error ? `${errorMessage}: ${error.message}` : `${errorMessage}: Unknown error`,
        "runtime_error"
      ));
      this.processNextInQueue(worker, mode);
    }
  }
  // Get the current queue size
  get queueSize() {
    return this._queue.length;
  }
  // Check if generation is in progress
  get isGenerating() {
    return this._isGenerating;
  }
  // Check if there's a current execution (useful for debugging)
  get hasCurrentExecution() {
    return this._currentExecution !== null;
  }
  // Get current execution type (useful for debugging)
  get currentExecutionType() {
    if (!this._currentExecution) return null;
    return this._currentExecution.isStream ? "generateStream" : "generate";
  }
  // Clear the queue
  clearQueue() {
    this._queue = [];
  }
  // Force clear everything including current execution (emergency use)
  forceReset() {
    this._queue = [];
    this._currentExecution = null;
    this._isGenerating = false;
  }
};

// src/managers/webai-worker-manager.ts
var WorkerManager = class {
  _worker = null;
  _workerPath;
  _modelId;
  _cachedWorkerScript = null;
  constructor(workerPath, modelId) {
    this._workerPath = workerPath;
    this._modelId = modelId;
  }
  async initWorker() {
    try {
      if (!this._cachedWorkerScript) {
        if (isLocalPath(this._workerPath)) {
          this._worker = new Worker(this._workerPath, {
            type: "module"
          });
          if (!this._worker) {
            throw createWorkerError(
              "Failed to initialize WebAI worker from local path.",
              "initialization_failed"
            );
          }
          await this.setupWorkerListeners();
          return;
        }
        const workerPath = this._workerPath + "?t=" + Date.now();
        const workerFetchResponse = await fetch(workerPath, {
          cache: "no-store"
        });
        if (!workerFetchResponse.ok) {
          throw createWorkerError(
            `Failed to fetch worker script: ${workerFetchResponse.status} ${workerFetchResponse.statusText}`,
            "initialization_failed"
          );
        }
        this._cachedWorkerScript = await workerFetchResponse.text();
        if (!this._cachedWorkerScript) {
          throw new Error("Failed to load worker script");
        }
      }
      const workerBlob = new Blob([this._cachedWorkerScript], {
        type: "application/javascript"
      });
      const objectURL = URL.createObjectURL(workerBlob);
      this._worker = new Worker(objectURL, {
        type: "module"
      });
      URL.revokeObjectURL(objectURL);
      if (!this._worker) {
        throw createWorkerError(
          "Failed to initialize WebAI worker.",
          "initialization_failed"
        );
      }
      await this.setupWorkerListeners();
    } catch (error) {
      if (this._worker) {
        this._worker.terminate();
        this._worker = null;
      }
      if (isWorkerError(error)) {
        throw error;
      } else {
        if (error instanceof Error) {
          throw createWorkerError(
            `Failed to initialize worker: ${error.message}`,
            "initialization_failed"
          );
        } else {
          throw createWorkerError(
            "Failed to initialize worker: Unknown error",
            "initialization_failed"
          );
        }
      }
    }
  }
  async reloadWorker() {
    console.log("Terminating existing worker...");
    if (this._worker) {
      this._worker.terminate();
      this._worker = null;
    }
    await new Promise((resolve) => setTimeout(resolve, 100));
    console.log("Initializing new worker...");
    await this.initWorker();
    if (!this._worker) {
      throw createWorkerError(
        "Failed to create new worker instance after reload",
        "initialization_failed"
      );
    }
    console.log("Worker successfully reloaded");
  }
  async setupWorkerListeners() {
    return new Promise((resolve, reject) => {
      const handleError = (event) => {
        this._worker?.removeEventListener("error", handleError);
        this._worker?.removeEventListener("messageerror", handleMessageError);
        this._worker?.removeEventListener("message", handleMessage);
        reject(createWorkerError(
          `WebAI Worker runtime error: ${event.message || "Unknown error"}`,
          "runtime_error"
        ));
      };
      const handleMessageError = (event) => {
        this._worker?.removeEventListener("error", handleError);
        this._worker?.removeEventListener("messageerror", handleMessageError);
        this._worker?.removeEventListener("message", handleMessage);
        reject(createWorkerError(
          "Failed to deserialize message from worker",
          "message_error"
        ));
      };
      const handleMessage = (event) => {
        if (event.data?.type === "error") {
          this._worker?.removeEventListener("error", handleError);
          this._worker?.removeEventListener("messageerror", handleMessageError);
          this._worker?.removeEventListener("message", handleMessage);
          const workerError = this.parseWorkerError(event.data.data);
          console.log(`WebAI Worker initialization error: ${workerError.message}`);
          reject(workerError);
          return;
        }
        if (event.data?.type === "worker initialized" && event.data.data.success) {
          console.log(`WebAI worker successfully initialized!`);
          this._worker?.removeEventListener("error", handleError);
          this._worker?.removeEventListener("messageerror", handleMessageError);
          this._worker?.removeEventListener("message", handleMessage);
          resolve();
        }
      };
      this._worker?.addEventListener("error", handleError);
      this._worker?.addEventListener("messageerror", handleMessageError);
      this._worker?.addEventListener("message", handleMessage);
    });
  }
  /**
   * Parse worker error data and create appropriate error with context
   */
  parseWorkerError(errorData) {
    const errorMessage = errorData?.message || "Unknown worker error";
    const errorContext = errorData?.context || "unknown";
    return createWorkerError(
      `${errorContext}: ${errorMessage}`,
      "runtime_error"
    );
  }
  /**
   * Creates a message handler function that handles both success and error cases
   */
  createMessageHandler(expectedType, onSuccess, onError, cleanup) {
    return (event) => {
      if (event.data.type === "error") {
        cleanup();
        const workerError = this.parseWorkerError(event.data.data);
        console.log(`WebAI Worker error: ${workerError.message}`);
        onError(workerError);
        return;
      }
      if (event.data.type === expectedType) {
        cleanup();
        onSuccess(event.data.data);
      }
    };
  }
  /**
   * Creates a promise-based worker operation handler
   */
  createWorkerPromise(messageType, messageData, expectedResponseType) {
    return new Promise((resolve, reject) => {
      if (!this._worker) {
        reject(createWorkerError(
          "Worker not initialized",
          "initialization_failed"
        ));
        return;
      }
      const cleanup = () => {
        this._worker?.removeEventListener("message", handler);
      };
      const handler = this.createMessageHandler(
        expectedResponseType,
        (data) => resolve(data),
        (error) => reject(error),
        cleanup
      );
      this._worker.addEventListener("message", handler);
      this._worker.postMessage({ type: messageType, data: messageData });
    });
  }
  /**
   * Helper method for operations that expect generic success response
   */
  async sendWorkerMessage(messageType, messageData) {
    return this.createWorkerPromise(
      messageType,
      messageData,
      messageType
      // Expect response type to match message type
    );
  }
  terminateWorker() {
    if (this._worker) {
      this._worker.terminate();
      this._worker = null;
      console.log("WebAI worker terminated successfully.");
    }
  }
  interruptWorker() {
    if (this._worker) {
      this._worker.postMessage({ type: "interrupt" });
    }
  }
  async terminateAndReload() {
    if (this._worker) {
      this._worker.terminate();
    }
    this._worker = null;
    await this.initWorker();
  }
  get isWorkerActive() {
    return this._worker !== null;
  }
  clearMemory() {
    if (this._worker) {
      this._worker.postMessage({ type: "clearMemory" });
    }
  }
  /**
   * Clear the cached script (useful if you need to force re-fetch)
   */
  clearCache() {
    this._cachedWorkerScript = null;
  }
  get worker() {
    return this._worker;
  }
};

// src/webai.ts
var WebAI = class _WebAI {
  version = VERSION;
  _modelId;
  _dev;
  _workerPath;
  _onAuth;
  _authRetryOptions;
  _isWebGPUAvailable = null;
  _storageQuota = null;
  _isInitialized = false;
  // Managers
  _workerManager;
  _queueManager;
  _modelManager;
  constructor({
    modelId,
    dev = false,
    workerPath,
    onAuth,
    authRetryOptions
  }) {
    this._modelId = modelId;
    this._dev = dev;
    this._workerPath = workerPath;
    this._onAuth = onAuth;
    this._authRetryOptions = authRetryOptions;
    this._workerManager = new WorkerManager(this.#getWorkerPath(), modelId);
    this._queueManager = new QueueManager();
    this._modelManager = new ModelManager(modelId, this._workerManager, dev, onAuth, authRetryOptions);
  }
  static async create({
    modelId,
    dev = false,
    workerPath,
    onAuth,
    authRetryOptions
  }) {
    const instance = new _WebAI({ modelId, dev, workerPath, onAuth, authRetryOptions });
    await instance.#createInstance();
    return instance;
  }
  async #createInstance() {
    try {
      await this._workerManager.initWorker();
      this._isWebGPUAvailable = await checkIsWebGPUAvailable();
      this._storageQuota = await checkStorageQuota();
      await this._modelManager.checkModelSupports(this._workerManager.worker);
    } catch (error) {
      throw error;
    }
  }
  async init({
    mode = "auto",
    precision,
    device,
    priorities,
    onDownloadProgress,
    callbackThrottle = 1e3
  }) {
    const result = await this._modelManager.initModel({
      mode,
      precision,
      device,
      priorities,
      onDownloadProgress,
      callbackThrottle,
      worker: this._workerManager.worker,
      isWebGPUAvailable: this._isWebGPUAvailable,
      storageQuota: this._storageQuota
    });
    this._isInitialized = result;
    return result;
  }
  async generate(data) {
    if (!this._isInitialized) {
      throw new Error("WebAI instance must be initialized before generating. Call init() first.");
    }
    return this._queueManager.enqueueGenerate(
      data,
      this._workerManager.worker,
      this._modelManager.mode
    );
  }
  async generateStream(data) {
    if (!this._isInitialized) {
      throw new Error("WebAI instance must be initialized before generating. Call init() first.");
    }
    return this._queueManager.enqueueGenerateStream(
      data,
      this._workerManager.worker,
      this._modelManager.mode
    );
  }
  clearQueue({ interrupt = true } = {}) {
    this._queueManager.clearQueue();
    if (interrupt) {
      this.interrupt({ clearQueue: false });
    }
  }
  async deleteDownloadedModel({ precision } = {}) {
    if (precision) {
      await this._modelManager.deletePrecisionSpecificModel(precision);
    } else {
      await this._modelManager.clearCache();
    }
  }
  terminate() {
    this.clearQueue();
    this._workerManager.terminateWorker();
    this._isInitialized = false;
  }
  async interrupt({ clearQueue = true } = {}) {
    if (clearQueue) {
      this._queueManager.clearQueue();
    }
    if (this._modelManager.device === "wasm" || this._modelManager.externalInterrupt) {
      await this._handleWasmInterrupt();
    } else {
      this._workerManager.interruptWorker();
    }
  }
  async _handleWasmInterrupt() {
    try {
      this._queueManager.handleForcedInterruption();
      await this._workerManager.terminateAndReload();
      if (this._isInitialized && this._modelManager.mode) {
        await this._modelManager.reinitializeAfterInterrupt(
          this._workerManager.worker
        );
      }
    } catch (error) {
      console.error("Error during interrupt:", error);
    }
  }
  clearMemory() {
    this._workerManager.clearMemory();
    this._isInitialized = false;
  }
  async checkIsModelDownloaded({ precision }) {
    const modelKeys = this._modelManager.modelSupportedPrecisionsDevicesMap?.[precision]?.modelKeys;
    if (!modelKeys) {
      console.warn("Model does not support this precision", precision);
      return false;
    }
    const isDownloaded = await checkIsModelDownloaded(
      precision,
      modelKeys,
      this._modelId
    );
    return isDownloaded;
  }
  async downloadModel({
    precision,
    onDownloadProgress,
    callbackThrottle = 1e3
  }) {
    return this._modelManager.downloadModel(
      this._workerManager.worker,
      precision,
      onDownloadProgress,
      callbackThrottle
    );
  }
  #getWorkerPath() {
    if (this._workerPath) {
      return this._workerPath;
    }
    const basePath = `${REMOTE_WORKER_ENDPOINT}/models/${this._modelId}/js-worker`;
    const workerFile = `${this._modelId}.worker.js`;
    return `${basePath}/${workerFile}`;
  }
  get modelId() {
    return this._modelId;
  }
  get dev() {
    return this._dev;
  }
  get mode() {
    return this._modelManager.mode;
  }
  get precision() {
    return this._modelManager.precision;
  }
  get device() {
    return this._modelManager.device;
  }
  get isWebGPUAvailable() {
    return this._isWebGPUAvailable;
  }
  get deviceStorageQuota() {
    return this._storageQuota;
  }
  get modelSupportedPrecisionsDevicesMap() {
    return this._modelManager.modelSupportedPrecisionsDevicesMap;
  }
  get modelSupportedPrecisions() {
    return this._modelManager.modelSupportedPrecisions;
  }
  get modelSupportedModes() {
    return this._modelManager.modelSupportedModes;
  }
  get isGenerating() {
    return this._queueManager.isGenerating;
  }
  get queueSize() {
    return this._queueManager.queueSize;
  }
  get doesSupportStreamGeneration() {
    return this._modelManager.doesSupportStreamGeneration;
  }
  get isInitialized() {
    return this._isInitialized;
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ErrorCategory,
  WebAI,
  checkIsModelDownloaded,
  checkIsWebGPUAvailable,
  checkStorageQuota,
  clearModelCache,
  createAuthError,
  createError,
  createNetworkError,
  createWorkerError,
  identifyError,
  isAuthError,
  isErrorOf,
  isLocalPath,
  isNetworkError,
  isWorkerError
});
